### PR TITLE
Expose screen dimensions

### PR DESCRIPTION
**Pull request by skamensky**
PR for issue https://github.com/wailsapp/wails/issues/1518

*Status*: working for linux
*Upcoming changes*: macos and windows implementation of get dimensions
*Requested feedback* before I continue working on this:

1. api. Should this be different?
2. Does scale need to be handled or can we just return the height and width as it is from gtk? From the docs it seems this is fine and it also seems to be how it's used for other apis in wails
3. Did I modify all of the correct places for adding a new window call? If not, is there a comprehensive list of places to change? Are any of these auto-gererated
4. Should this even be called window or should be add a new "namespace" of screen since technically the api is unrelated to any individual window
5. Do we have somewhere to add tests?
6. General feedback (this is my first time using golang)

*Some other questions*

**When do we put c implementation in the  `/v2/internal/ffenestri/ffenestri_{platform}.c` files and when do we put them in the `/v2/internal/frontend/desktop/{platform}/window.go` files?**

**When do we need wait groups+invokeOnMainThread?**
for example I see a wait group here 

https://github.com/wailsapp/wails/blob/41d1bf3d109a002b7dbb45cb27dd3cdfac2f38cf/v2/internal/frontend/desktop/linux/window.go#L709-L719

But none here

https://github.com/wailsapp/wails/blob/41d1bf3d109a002b7dbb45cb27dd3cdfac2f38cf/v2/internal/frontend/desktop/linux/window.go#L733-L737

**When do we need to add to this switch statement?**

https://github.com/wailsapp/wails/blob/41d1bf3d109a002b7dbb45cb27dd3cdfac2f38cf/v2/internal/frontend/dispatcher/systemcalls.go#L23-L41

### CODE REVIEW

File: v2/internal/frontend/runtime/wrapper/runtime.d.ts [hunk 1]
```diff
@@ -131,6 +131,10 @@ export function WindowSetMaxSize(width: number, height: number): void;
 // Setting a size of 0,0 will disable this constraint.
 export function WindowSetMinSize(width: number, height: number): void;
 
+// [WindowGetDimensions](https://wails.io/docs/reference/runtime/window#windowgetdimensions)
+// Gets the dimensions of the screen.
+export function WindowGetSize(): Promise<Size>;
```
Comment (leaanthony) [inline v2/internal/frontend/runtime/wrapper/runtime.d.ts:136]: Should this be `WindowGetDimensions`?

**Comment by leaanthony** [comment]
Thanks for this! 🚀 

To answer your questions:

> api. Should this be different?

I think it's ok. The thing to keep in mind sometimes people have multiple monitors, which will affect the API.

> Does scale need to be handled or can we just return the height and width as it is from gtk? From the docs it seems this is fine and it also seems to be how it's used for other apis in wails

Do you mean the DPI of the screens? This is handled behind the scenes for most scenarios. I think it's ok not to report that in this instance.

> Did I modify all of the correct places for adding a new window call? If not, is there a comprehensive list of places to change? Are any of these auto-gererated

I'll add more documentation on how to do this and use this as an example for that. The only places you've left out are the darwin+windows implementations of the method but I'm assuming that's deliberate for now. These aren't auto-generated but that's not a bad idea ;-)

> Should this even be called window or should be add a new "namespace" of screen since technically the api is unrelated to any individual window

I think I agree with you that it needs to be in a new "namespace". You suggested "Screen" and that's not a bad option. 

> Do we have somewhere to add tests?

Testing is *hard* for gui apps, but unit tests can be added using the `<filename>_test.go` convension. You can find some examples of this in the repo.

> General feedback (this is my first time using golang)

You're doing well - keep going! 👍

**Comment by skamensky** [comment]
Thank you for the feedback. I'm working on this. Will post back when I have something working for Linux with a new 'screen' api.

File: v2/internal/frontend/runtime/wrapper/runtime.d.ts (hunk 1 above)
Comment (skamensky) [reply v2/internal/frontend/runtime/wrapper/runtime.d.ts:136]: Yup, my mistake. Will fix this in the next public commit

**Comment by skamensky** [comment]
@leaanthony , This PR now exposes a `Screen` api. As of now it only exposes a `ScreenGetAll` function but convenience functions can be added if needed.

Here is some sample go and js on my system (with outputs)

```golang
func (a *App) OnDomReady(ctx context.Context) {
	screens, err := runtime.ScreenGetAll(ctx)
	if err != nil {
		fmt.Printf("Error getting screens: %v\n", err)
		return
	}
	for _, screen := range screens {
		fmt.Printf("{height=%v, width=%v IsCurrent=%v, IsPrimary=%v}\n", screen.Height, screen.Width, screen.IsCurrent, screen.IsPrimary)
	}
}
```

*Output*
```
{height=720, width=1280 IsCurrent=true, IsPrimary=true}
{height=540, width=960 IsCurrent=false, IsPrimary=false}
```

```js
runtime.ScreenGetAll().then(screens=>{
    screens.forEach(screen=>{
        console.log(screen)
    })    
})
```

*Output*

```js
{isCurrent: true, isPrimary: true, width: 1280, height: 720}
{isCurrent: false, isPrimary: false, width: 960, height: 540}
```

Is this API OK now? Should I get started on Windows/MacOS, or are there some changes/a different that you'd like to see?

File: v2/internal/frontend/desktop/linux/window.go [hunk 2]
```diff
@@ -48,7 +48,7 @@ static void SetMinMaxSize(GtkWindow* window, int min_width, int min_height, int
 GdkMonitor* getCurrentMonitor(GtkWindow *window) {
 	// Get the monitor that the window is currently on
 	GdkDisplay *display = gtk_widget_get_display(GTK_WIDGET(window));
-	GdkWindow *gdk_window = gtk_widget_get_window(GTK_WIDGET(window));
```
Comment (skamensky) [inline v2/internal/frontend/desktop/linux/window.go:51]: This should be removed from this PR and put into a different PR and a different commit. In fact, this isn't even the change I meant to introduce (however, there is a bug in this code that I have a fix for).

**Comment by leaanthony** [comment]
Looking good so far 👍

**Comment by skamensky** [comment]
Glad to report that the windows implementation is working!

As an aside @leaanthony how have you debugged on windows? I couldn't get the windows commant prompt or powershell to output to stdout, it simply opens the windows which is another process. I attmpeted to use procmon and windows debugger but I couldn't figure out how to view simple `fmt.Printf` outputs.

windows_screen_api_working

**Comment by leaanthony** [comment]
I just use the `-ide goland` flag when generating the project and use the ide debugger

**Comment by skamensky** [comment]
@leaanthony , all three platforms have been implemented. What should the next steps be?

**Comment by leaanthony** [comment]
Need to find time to review it 👍

**Review by leaanthony — CHANGES_REQUESTED** [review]
This is looking and working great! My only comment is that on MacOS, the screen height looks to be affected if the task bar is there. If I'm on a screen with a taskbar, it reports 64 pixels *less* than when the taskbar is hidden. I think this is because the code is using `visibleFrame` instead of [`frame`](https://developer.apple.com/documentation/appkit/nsscreen/1388387-frame?language=objc). So we could either return both, or keep it consistent between the platforms (report full screen size for all platforms). Thoughts?

File: v2/internal/frontend/desktop/windows/screen.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	return first.DwFlags == second.DwFlags &&
+		first.RcMonitor.Top == second.RcMonitor.Top &&
+		first.RcMonitor.Bottom == second.RcMonitor.Bottom &&
+		first.RcMonitor.Right == second.RcMonitor.Right &&
+		first.RcMonitor.Left == second.RcMonitor.Left &&
+		first.RcWork.Top == second.RcWork.Top &&
+		first.RcWork.Bottom == second.RcWork.Bottom &&
+		first.RcWork.Right == second.RcWork.Right &&
+		first.RcWork.Left == second.RcWork.Left &&
+		first.DwFlags == second.DwFlags
```
[the remark this answers is not in the document]
Comment (skamensky) [reply v2/internal/frontend/desktop/windows/screen.go:26]: I didn't know if comparing `RcMonitor` only instead of also comparing `RcWork` was enough to guarantee uniqueness of a monitor. Does `RcMonitor` carry enough uniqueness to drop the comparison with `RcWork`

**Comment by skamensky** [comment]
https://github.com/wailsapp/wails/pull/1519#pullrequestreview-1038736485
Yes, I wasn't sure what the difference was, now that I understand currectly, we should use frame instead of viewframe. I agree, the meaning of `Screen` should stay consistent across platforms.


Overall I would like to get your feedback on the rest of the questions I left i response and clean the code up a bit (I think I have a typo in a function name somewhere, I need to find it).

Should squash the commits that I think don't contribute anything? Or would you prefer I leave the history as is?

File: v2/internal/frontend/runtime/desktop/screen.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+
+import {Call} from "./calls";
+
+
+/**
+ * // Gets the all screens. Call this anew each time you want to refresh data from the underlying windowing system.
+ *
+ * @export
+ * @return {Promise<{isCurrent: boolean; isPrimary: boolean; width : number  height : number}>} The screens
```
[the remark this answers is not in the document]
Comment (leaanthony) [reply v2/internal/frontend/runtime/desktop/screen.js:21]: JSDoc, though I think you can use typescript syntax.

File: v2/internal/frontend/desktop/windows/screen.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (leaanthony) [reply v2/internal/frontend/desktop/windows/screen.go:26]: Sorry, I meant you have `first.DwFlags == second.DwFlags` twice

[the remark this answers is not in the document]
Comment (skamensky) [reply v2/internal/frontend/desktop/windows/screen.go:26]: Ah, I missed that. Yes, I'll remove that.

**Comment by skamensky** [comment]
@leaanthony ,

I've taken care of all of the changes requested.

Anything else you'd like me to do here?

BTW, this question still stands:


> Should squash the commits that I think don't contribute anything? Or would you prefer I leave the history as is?

**Comment by leaanthony** [comment]
I'll check it out over the weekend. We'll squash merge so all good 👍

**Comment by leaanthony** [comment]
@all-contributors please add @skamensky  for code, ideas and docs 👍

**Comment by leaanthony** [comment]
Nice one!

**Comment by skamensky** [comment]
Yay thank you for guiding me through my first open source contribution!

**Comment by leaanthony** [comment]
First? Really? You did a fantastic job! 👏
